src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,366 @@
+                f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+            )
+            status_code = comp_data.status_code
+            if status_code == 200:
+                comp_data = comp_data.json()
+                auction_id = comp_data["auctionId"]
+                bucket_response = dict(
+                    requests.get(
+                        f"https://solver-instances.s3.eu-central-1.amazonaws.com/prod-mainnet/{auction_id}.json"
+                    ).json()
+                )
+            elif status_code == 404:
+                # attempt fetch from staging environment, if prod failed.
+                comp_data = requests.get(
+                    f"https://barn.api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+                )
+                status_code = comp_data.status_code
+                if comp_data.status_code == 200:
+                    comp_data = comp_data.json()
+                    auction_id = comp_data["auctionId"]
+                    bucket_response = dict(
+                        requests.get(
+                            f"https://solver-instances.s3.eu-central-1.amazonaws.com/staging-mainnet/{auction_id}.json"
+                        ).json()
+                    )
+            return comp_data["solutions"][-1]["orders"], bucket_response
+        except ValueError as except_err:
+            self.logger.error("Unhandled exception: %s", str(except_err))
+
+    def get_win_trade_surplus(self, decoded_settlement, trade, order_type):
+        """
+        Extracts necessary data from trade, and clearing prices of on-chain settlement.
+        Makes call to surplus calculating function.
+        """
+        sell_token_index = trade["sellTokenIndex"]
+        buy_token_index = trade["buyTokenIndex"]
+        sell_token_clearing_price = decoded_settlement.clearing_prices[sell_token_index]
+        buy_token_clearing_price = decoded_settlement.clearing_prices[buy_token_index]
+        return self.get_surplus(
+            trade, sell_token_clearing_price, buy_token_clearing_price, order_type
+        )
+
+
+    def get_solver_trade_surplus(self, trade, order, instance2, order_type):
+        """
+        Extracts necessary data from trade, and clearing prices returned by Quasimodo.
+        Makes call to surplus calculating function.
+        """
+        sell_token = order["sell_token"]
+        buy_token = order["buy_token"]
+        sell_token_clearing_price = instance2["prices"][sell_token]
+        buy_token_clearing_price = instance2["prices"][buy_token]
+
+        return self.get_surplus(
+            trade, sell_token_clearing_price, buy_token_clearing_price, order_type
+        )
+
+
+    def decode_single_hash(self, settlement_hash):
+        """
+        Need a better name for this function. Goes through all settlements fetched, decodes orders,
+        gets response from AWS bucket and makes call to Quasimodo for required orders.
```

Comment:
```suggestion
        gets response from AWS bucket and makes call to Quasimodo for required orders.
        Returns `None` in case data cannot be fetched, `"200"` otherwise.
```
(Not the cleanest approach to return values, but the comment will make it easier to change stuff later.)

Reply:
We can still make it clean.
Is it because of the 200? Basically, we just want to return something not None if order is parsed and computed by quasimodo successfully. We can return True too.